File: ref/code/L09/convention2.java
package com.deltacoding.usaco.s2018_2019.dec2018.silver;

import java.io.*;
import java.util.*;

public class convention2 {

	public static void main(String[] args) throws IOException {

		BufferedReader in = new BufferedReader(new FileReader("convention2.in"));
		PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("convention2.out")));

		int n = Integer.parseInt(in.readLine());
		Cow[] cows = new Cow[n];

		for (int i = 0; i < n; i++) {
			StringTokenizer tk = new StringTokenizer(in.readLine());
            int arrive = Integer.parseInt(tk.nextToken());
            int eating = Integer.parseInt(tk.nextToken());
			cows[i] = new Cow(arrive, eating, i);
		}

		Arrays.sort(cows, new Comparator<Cow>() {
			@Override
			public int compare(Cow c1, Cow c2) {
				if(c1.arrive != c2.arrive) {
					return Integer.compare(c1.arrive, c2.arrive);
				}
				return Integer.compare(c1.seniority, c2.seniority); // 不要漏掉
			}
		});

		PriorityQueue<Cow> q = new PriorityQueue<Cow>(new Comparator<Cow>() {
			@Override
			public int compare(Cow c1, Cow c2) {
				return Integer.compare(c1.seniority, c2.seniority);
			}
		});

		int answer = 0;
		int last = cows[0].arrive; // 上一头奶牛吃草结束时间
		q.offer(cows[0]);
		int cnt = 1; // 已加入到优先队列中的牛

		while (!q.isEmpty()) {
			Cow c = q.poll();
			int wait = Math.max(0, last - c.arrive);

			answer = Math.max(answer, wait);

			last = c.arrive + wait + c.t; // 当前吃草的牛结束时间

			for (int i = cnt; i < n; i++) {
				if (cows[i].arrive <= last) { // 当前吃草的牛结束时，把已经到达的牛加入队列
					q.offer(cows[i]);
					cnt++;
				} else {
					break;
				}
			}

			if (q.isEmpty() && cnt < n) { // 不能漏
				q.offer(cows[cnt]);
				cnt++;
			}
		}

		out.println(answer);
		in.close();
		out.close();

	}

	static class Cow {
		int arrive;
		int t;
		int seniority;

		public Cow(int a, int b, int c) {
			arrive = a;
			t = b;
			seniority = c;
		}
	}
}

File: ref/code/L09/convention2.cpp
#include <bits/stdc++.h>
using namespace std;

struct cow {
    int arrive, t, seniority;
};

cow cows[100001];

bool cmp(const cow &c1, const cow &c2) {
    if (c1.arrive != c2.arrive) {
        return c1.arrive < c2.arrive;
    }
    return c1.seniority < c2.seniority;  // 不要漏掉
}

struct cmpPrior {
    bool operator()(const cow &a, const cow &b) {
        return a.seniority > b.seniority;
    }
};

priority_queue<cow, vector<cow>, cmpPrior> pq; // 当前正在等待吃草的牛

int main() {
    freopen("convention2.in", "r", stdin);
    freopen("convention2.out", "w", stdout);
    int n;
    cin >> n;
    for (int i = 0; i < n; i++) {
        int a, t;
        scanf("%d%d", &a, &t);
        cows[i] = {a, t, i};
    }
    sort(cows, cows + n, cmp);

    pq.push(cows[0]);
    int last = cows[0].arrive;  // 上一头牛吃草结束时间
    int answer = 0;
    int cnt = 1;  // 已加入队列中的牛头数

    while (!pq.empty()) {
        cow c = pq.top();
        pq.pop();
        int wait = max(0, last - c.arrive);
        answer = max(answer, wait);
        last = c.arrive + wait + c.t;
        // 把已经到达的牛加入队列
        for (int i = cnt; i < n; i++) {
            if (cows[i].arrive <= last) {
                pq.push(cows[i]);
                cnt++;
            } else {
                break;
            }
        }
        // 当队列中为空，表示当前这一头牛吃草结束时，后续的牛还未到达
        if (pq.empty() && cnt < n) {
            pq.push(cows[cnt]);
            cnt++;
        }
    }
    cout << answer << endl;

    return 0;
}
